fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -3812,8 +3812,7 @@ class FileBackupAgentImpl {
 		Optional<Value> lastBackupTimestamp = wait(backupAgent->lastBackupTimestamp().get(tr));
 
 		if ((lastBackupTimestamp.present()) && (lastBackupTimestamp.get() >= nowStr)) {
-			fprintf(stderr, "ERROR: The last backup `%s' happened in the future.\n", printable(lastBackupTimestamp.get()).c_str());
-			throw backup_error();
+			fprintf(stderr, "WARNING: The last backup `%s' appears to have started in the future.\n", printable(lastBackupTimestamp.get()).c_str());
```

Comment:
Are there any weird implications of equality with a prior timestamp?

Reply:
There is one indirect case.  If the user is backing up to a local filesystem and using a backup naming convention such that two backups with the same timestamp will have the same name, then this would cause two backups to use the same destination folder.

Logic around backup to local filesystems is tricky because FDB does not require that all hosts running `fdbbackup` or the backup agents (or `fdbserver` in the new log mode in 6.3) see the same filesystem.  Therefore it is valid for the submitting host to not have access to the destination path, and for each agent to be writing to a different filesystem at that same path, and then the user must combine the contents of all of those separate locations later to perform a restore.  Because of this, creating the destination directory from the submitting host might not be possible, so this isn't done and it isn't possible to prevent reuse of a destination from the submitting host.

Perhaps an error should still be thrown in case that the last backup timestamp is equal to now.

Reply:
The error might be helpful, but it's interesting to note that I don't think it actually catches all of the cases. For example, it could happen that the last timestamp is not the largest ever used and your current timestamp could have a collision even if it's larger than the last one.

In the case that you are smaller than the last backup timestamp, it's also clearly possible that some other backup in the past used the same value.

Reply:
Correct, but I think the general problem of destination folder re-use is not easily solved due to the weak requirements on accessibility of the destination folder.

Even if a cluster maintained a history of previously used backup timestamps, other clusters could write to the same backup name with the same timestamp if the other parts of the backup name were also the same.

While I would like to change the contract for `file://` backup destinations so that

- `fdbbackup` on the submitting host is responsible for creating the target directory, and its prior existence is an error
- `backup_agent` and `fdbserver` processes writing to the backup container expect the target directory to exist, and its absence is an error

which would prevent destination folder re-use, I know that there are people using local path backup destinations to write to non-shared filesystems so they rely on the current behavior.

Reply:
I don't necessarily have a problem with not resolving the collision risk in this PR, but I want to make sure that the outcome of a collision isn't so bad that we are making things worse by performing this change.

If we do go ahead with this PR, then it may be worth adding a comment to the error check saying that it's more of an optimization rather than a full prevention of collisions.

Reply:
Reusing an output directory is pretty bad.  While it is a forensically recoverable situation, as-is the backup/restore tools will not be able to make sense of the directory contents.

As of now, a repeat exact timestamp vs the last used timestamp will still throw an error.   I don't think the inequality change makes collision significantly more likely.  

We could, however, leave the original error check and throw in place, and the only side effect would be that if you start a backup in a timezone N hours ahead of UTC and then within N hours you update FDB and start a new backup then you would get the error that the last backup appears to be in the future.

Reply:
Summarizing this issue:

A regressing backup start timestamp string can cause a collision between an old backup's timestamped output folder and the new one.  This is bad and makes both backups unusable without recovery tools that do not currently exist.  

A collision with _exactly the most recently started backup timestamp_ will be detected, but a collision with a backup start timestamp prior to that would not be.  The latter is new behavior, as the previous logic would not allow any backup start timestamp regression, each new backup start timestamp had to be greater than the last.

Thinking about what, in practice, would cause a collision between non-consecutive backups, it occurs to me that an automated task that runs a non-continuous full backup every hour on-the-hour would generate timestamps that are typically exactly 1 hour apart.  Therefore, if the timestamp string used by backup were shifted back by multiple timezones a collision is almost certain.  

Currently, such a shift could be caused by starting a backup from a different timezone than previously, and it would result in an error.  After the change to use UTC, such a shift _will_ occur for all backups started from timezones 1 hour or more greater than GMT, which seems dangerous.

For this reason, I will change the logic back to always throw an error if the backup start timestamp string regresses.

Unfortunately, this does mean that to avoid being unable to start a backup due to this issue one most wait N hours after starting a backup before upgrading their FDB cluster to a version that includes this change, where N is the number of hours greater than GMT the last backup was started in.

@sfc-gh-ajbeamon Does this sound good?   I'm not sure how best to communicate the caveat above.

Reply:
Is it possible that we could avoid collisions by using an output directory that included a random ID in addition to the timestamp?

If not and there's no other way to avoid directory reuse, then it sounds reasonable to default to the safe behavior and disallow starting the backup. In we do this, we could potentially provide a `--force` mechanic to let people override it if they felt their situation warranted it.